File: frontend/src/services/cacheService.js
/**
 * Simple cache service for power readings data
 */
class CacheService {
    constructor() {
        this.cache = {};
    }
    
    /**
     * Get cache key for node, year, month, day combination
     * @param {string} node - Node ID
     * @param {string} year - Year
     * @param {string} month - Month
     * @param {string} day - Day
     * @returns {string} Cache key
     */
    getCacheKey(node, year, month, day) {
        return `${node}_${year}_${month}_${day}`;
    }
    
    /**
     * Check if data exists in cache
     * @param {string} node - Node ID
     * @param {string} year - Year
     * @param {string} month - Month
     * @param {string} day - Day
     * @returns {Array|null} Cached data or null
     */
    checkCache(node, year, month, day) {
        const key = this.getCacheKey(node, year, month, day);
        return this.cache[key] || null;
    }
    
    /**
     * Add data to cache
     * @param {string} node - Node ID
     * @param {string} year - Year
     * @param {string} month - Month
     * @param {string} day - Day
     * @param {Array} data - Data to cache
     */
    addToCache(node, year, month, day, data) {
        const key = this.getCacheKey(node, year, month, day);
        this.cache[key] = data;
    }
    
    /**
     * Clear cache for a specific node or all cache
     * @param {string} [node] - Optional node ID to clear
     */
    clearCache(node) {
        if (node) {
            // Clear only keys for this node
            Object.keys(this.cache).forEach(key => {
                if (key.startsWith(`${node}_`)) {
                    delete this.cache[key];
                }
            });
        } else {
            // Clear all cache
            this.cache = {};
        }
    }
    
    /**
     * Get all cached nodes
     * @returns {string[]} Array of node IDs with cached data
     */
    getCachedNodes() {
        const nodes = new Set();
        Object.keys(this.cache).forEach(key => {
            const nodePart = key.split('_')[0];
            nodes.add(nodePart);
        });
        return Array.from(nodes);
    }
}

// Create a singleton instance
const cacheService = new CacheService();
export default cacheService;

File: backend/power_monitor/services/cache_service.py
import time
from datetime import datetime, timedelta

class CacheService:
    _instance = None  # Singleton instance
    _cache = {}       # In-memory cache
    _cache_ttl = {}   # Cache expiration times
    
    def __new__(cls):
        """Ensures only one instance of CacheService exists."""
        if cls._instance is None:
            cls._instance = super(CacheService, cls).__new__(cls)
            cls._instance.initialize()
        return cls._instance
    
    def initialize(self):
        """Initialize the cache service."""
        self._cache = {}
        self._cache_ttl = {}
        self._default_ttl = 3600  # Default TTL: 1 hour (in seconds)
    
    def get_cache_key(self, node, year, month, day):
        """Generate consistent cache key."""
        return f"{node}_{year}_{month}_{day}"
    
    def get(self, node, year, month, day):
        """Get data from cache if it exists and hasn't expired."""
        key = self.get_cache_key(node, year, month, day)
        
        # Check if key exists in cache
        if key in self._cache:
            # Check if cache has expired
            if key in self._cache_ttl and self._cache_ttl[key] > time.time():
                print(f"Cache hit for {key}")
                return self._cache[key]
            else:
                # Cache expired, remove it
                print(f"Cache expired for {key}")
                if key in self._cache:
                    del self._cache[key]
                if key in self._cache_ttl:
                    del self._cache_ttl[key]
        
        print(f"Cache miss for {key}")
        return None
    
    def set(self, node, year, month, day, data, ttl=None):
        """Store data in cache with expiration time."""
        key = self.get_cache_key(node, year, month, day)
        
        # Store the data
        self._cache[key] = data
        
        # Set expiration time
        ttl = ttl or self._default_ttl
        self._cache_ttl[key] = time.time() + ttl
        
        print(f"Cached {len(data)} readings for {key}, expires in {ttl} seconds")
        return True
    
    def clear(self, node=None):
        """Clear all cache or just for a specific node."""
        if node:
            # Clear only keys belonging to the specified node
            keys_to_remove = [k for k in self._cache.keys() if k.startswith(f"{node}_")]
            for key in keys_to_remove:
                if key in self._cache:
                    del self._cache[key]
                if key in self._cache_ttl:
                    del self._cache_ttl[key]
            print(f"Cleared cache for node {node}")
        else:
            # Clear all cache
            self._cache = {}
            self._cache_ttl = {}
            print("Cleared all cache")
        
        return True
    
    def get_cached_nodes(self):
        """Get a list of nodes that have cached data."""
        # Extract unique node IDs from cache keys
        nodes = set()
        for key in self._cache.keys():
            parts = key.split('_')
            if len(parts) > 0:
                nodes.add(parts[0])
        
        return list(nodes)
    
    def get_cache_stats(self):
        """Get statistics about the cache."""
        # Count items by node
        node_counts = {}
        for key in self._cache.keys():
            parts = key.split('_')
            if len(parts) > 0:
                node = parts[0]
                if node in node_counts:
                    node_counts[node] += 1
                else:
                    node_counts[node] = 1
        
        total_items = len(self._cache)
        total_nodes = len(node_counts)
        
        # Calculate expiration
        now = time.time()
        active_ttls = [ttl - now for ttl in self._cache_ttl.values() if ttl > now]
        avg_ttl_remaining = sum(active_ttls) / len(active_ttls) if active_ttls else 0
        
        return {
            "total_cached_items": total_items,
            "total_nodes_cached": total_nodes,
            "items_by_node": node_counts,
            "avg_seconds_remaining": int(avg_ttl_remaining)
        }